utils/publish-snapshot.sh

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env bash
+
+# https://www.gnu.org/software/bash/manual/bash.html#The-Set-Builtin
+set -e # Exit immediately if a command exits with a non-zero status.
+set -o pipefail
+set -u # Treat unset variables as an error when substituting.
+
+# This script uses the `changeset` tool to version each package as a snapshot
+# release and then publishes that release to npm with a tag named after the PR.
+#
+# It is designed to be run in a Github Action, and by default will abort if the
+# `CI`, `GITHUB_EVENT_NAME`, or `GITHUB_REF` environment variables are unset.
+
+# Identifies the path that the script is in (http://stackoverflow.com/a/246128/11807)
+MYPATH=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
+
+# ROOT is the root directory of our project.
+ROOT="$MYPATH/.."
+
+pushd "$ROOT"
+
+if [ -z ${CI+CI_UNSET} ]; then
+    echo "CI environment variable is unset. Exiting!"
+    exit 1
+fi
+
+if [ -z ${GITHUB_EVENT_NAME+GITHUB_EVENT_NAME_UNSET} ]; then
+    echo "GITHUB_EVENT_NAME environment variable is unset. Exiting!"
+    exit 1
+fi
+
+if [ -z ${GITHUB_REF+GITHUB_REF_UNSET} ]; then
+    echo "GITHUB_REF environment variable is unset. Exiting!"
+    exit 1
+fi
+
+if [[
+    "$GITHUB_EVENT_NAME" != "workflow_dispatch" \
+    && "$GITHUB_EVENT_NAME" != "pull_request"
+]]; then
+    exit
+fi
+
+# Check if we need to do any work
+temp_file=$(mktemp)
+yarn changeset status --verbose --output "$temp_file"
+jq -e '.releases | length | if . > 0 then . else "Error: No changesets found\\n" | halt_error(1) end' < "$temp_file"
+
+
+echo "Running for $GITHUB_EVENT_NAME @ $GITHUB_REF"
+
+# Example GITHUB_REF
+# refs/pull/:prNumber/merge
+if [[ "$GITHUB_REF" =~ refs/pull/([[:digit:]]+)/merge ]]; then
+    echo "Found PR #${BASH_REMATCH[1]}"
+    PR_NUMBER="PR${BASH_REMATCH[1]}"
+else
+    echo "Pull Request number not found in ref. Exiting!"
+    exit 1
+fi
+
+# publish:ci steps
+node "$ROOT/utils/pre-publish-check-ci.js"
+
+if ! git diff --stat --exit-code HEAD; then
+    echo "Git repo is dirty. Stash/Commit your changes before running."
+    exit 1
+fi
+
+yarn build
+yarn extract-strings
+
+# Now version the packages and publish a snapshot
+yarn changeset version --snapshot "$PR_NUMBER"
+yarn changeset publish --no-git-tag --tag "${PR_NUMBER}"
```

Comment:
I'm not familiar with these options, what do they do?

Does npm allow you to republish using the same version?  What happens when someone updates a PR?

Reply:
Yeah, I'm curious about this to:
- Is it possible to publish to the same version (IIRC in Java I think you could if it was a snapshot, so maybe it's the same here)
- Are we worried about yarn's cache? If we request `perseus@PR404` twice expecting different results, will yarn just return the first instance from the cache each time?

I'm wondering if it would be safer to do `perseus@PR404-[SHA]` or `perseus@PR404-[TIMESTAMP]`.

---

I'm also thinking from the perspective of users updating versions in Webapp:
- Update Perseus version with snapshot to generate ZND
- Make changes to Perseus snapshot
- How do you trigger Webapp's CI/CD to regenerate the ZND? Since the version is the same, there's nothing to change, and without the change I'm wondering how we trigger CI/CD again.

Maybe this will be a moot point once we start autogenerating PRs in Webapp?

Reply:
> Does npm allow you to republish using the same version? 
Yes. What's happening here is that we're publishing under a tag. `npm` supports publishing a new version to a tag, in fact, when you do a regular publish, you are publishing to the `latest` tag and that bumps it to whatever version you published.  You can see this by running `yarn info --json react | jq '.data["dist-tags"]'`.  That will show you all the tags and the version associated with them. 

> What happens when someone updates a PR?
It will replace the tag for this PR with the newly published version.  

> Is it possible to publish to the same version (IIRC in Java I think you could if it was a snapshot, so maybe it's the same here)
No, but we're not publishing the same version, we're publishing to the same `npm` tag. Each new update to the PR will result in a new version for the PR's tag. 

> Are we worried about yarn's cache? If we request perseus@PR404 twice expecting different results, will yarn just return the first instance from the cache each time?
No. This is no different from publishing a regular npm package. If you look at `yarn info react` you'll see that there are a set of `dist-tags` and each points to a version. If you run `yarn add react@experimental`, you should always get the latest version of react published under the `experimental` tag. 

> I'm wondering if it would be safer to do perseus@PR404-[SHA] or perseus@PR404-[TIMESTAMP].
Under the hood, we _are_ publishing a version that is named like `@khanacademy/perseus@0.0.0-PR423-20230314222716`.  

> I'm also thinking from the perspective of users updating versions in Webapp
I _think_ we can "poke" the webapp PR and have it rebuild the ZND. Hmm... yeah, if we have nothing to change, how do we re-run the ZND build?  Technically, the `yarn.lock` file is never changed for these snapshot build changes so any build should automatically get the latest version of that tag. So we just need a way to "poke" the build to re-run. But maybe ti'd be easier to just use the explicit version created (see above). 

I think I'd like to leave things as they are with a tag. We _do_ get an explicit version for each PR update in Perseus so if we need to we can always "export" that version number and then use that to update the webapp PR.